src/v/s3/client.cc

```diff
@@ -734,11 +713,21 @@ ss::future<> client_pool::stop() {
 ss::future<client_pool::client_lease> client_pool::acquire() {
     gate_guard guard(_gate);
     try {
+        // If credentials have not yet been acquired, wait for them. It is
+        // possible that credentials are not initialized right after remote
+        // starts, and we have not had a response from the credentials API yet,
+        // but we have scheduled an upload. This wait ensures that when we call
+        // the storage API we have a set of valid credentials.
+        if (unlikely(!_apply_credentials)) {
+            co_await wait_for_credentials();
+        }
+
         while (_pool.empty() && !_gate.is_closed()) {
             if (_policy == client_pool_overdraft_policy::wait_if_empty) {
                 co_await _cvar.wait();
             } else {
-                auto cl = ss::make_shared<client>(_config, _as);
+                auto cl = ss::make_shared<client>(
+                  _config, _as, _apply_credentials);
```

Comment:
We don't ever clear `_apply_credentials` once it is set, we have it defined as a shared pointer to be able to lazy init the value because we depend on network calls to initialize it, as well as to share it with clients.

However, once we have initialized it, we only ever change the credential values it holds by calling `reset..` on the pointer - only the pool does the resets.

Reply:
> We don't ever clear _apply_credentials once it is set

I see. Then it may be sufficient to provide a well placed comment to explain this. Alternatively, you might consider deriving the initial value of _apply_credentials prior to the creation of the pool then its initial state would be non-null and I assume that would then eliminate the need for the condition variable etc..?

Reply:
I initially planned to make the credentials early in the lifecycle, but the chain of object creation is such that credentials (which are generated from network call) are always acquired late in the process:

1. application.cc - application creates cloud_storage_api (remote.cc) and starts the sharded service
2. cloud_storage_api - contains the client pool, creates it as part of its own constructor
3. cloud_storage_api - creates the background op, and passes a callback to it to regularly update the client pool with new credentials

The background op (refresh_credentials) calls the APIs to get credentials and updates the client pool, so we need the pool to be present before we can update it. 

Another option would be to do a blocking wait for the first set of credentials in the cloud_storage_api constructor, and then, once we have the credentials, create the client pool seeded with these credentials, and then launch the background op which would keep rotating credentials, but since the cloud_storage_api is constructed in application startup, I avoided blocking on the network call there.

If waiting for credentials is an option at startup though (potentially for several seconds), everything would be much simpler, as we can then create the pool after getting credentials as you mentioned.